numba/targets/arraymath.py

```diff
@@ -999,3 +1019,191 @@ def dot_3(context, builder, sig, args):
+
+    n = lda = a_shapes[0]
+
+    res = builder.call(fn, (kind_val, n,
+                            builder.bitcast(a_data, ll_void_p), lda,
+                            ipiv, builder.bitcast(work, ll_void_p),
+                            lwork, info
+                            ))
+    check_lapack_return(context, builder, res)
+
+
+def mat_inv(context, builder, sig, args):
+    """
+    Invert a matrix by first computing its LU decomposition.
+    """
+    xty = sig.args[0]
+    dtype = xty.dtype
+
+    x = make_array(xty)(context, builder, args[0])
+    x_shapes = cgutils.unpack_tuple(builder, x.shape)
+    m, n = x_shapes
+    check_c_int(context, builder, m)
+    check_c_int(context, builder, n)
+
+    # Allocate the return array (Numpy never works in place contrary to
+    # Scipy for which one can specify to whether or not to overwrite the
+    # input.
+    def create_out(a):
+        m, n= a.shape
+        if m != n:
+            raise ValueError("np.linalg.inv can only work on square "
+                             "arrays.")
+        return a.copy()
+
+    out = context.compile_internal(builder, create_out,
+                                   signature(xty, *sig.args), args)
+    o = make_array(xty)(context, builder, out)
+
+    # Allocate the array in which the pivot indices are stored.
+    ipiv_t = types.Array(types.intc, 1, 'C')
+    i = _empty_nd_impl(context, builder, ipiv_t, (m,))
+
+    info = cgutils.alloca_once(builder, ll_intc)
+
+    # Compute the LU decomposition of the matrix.
+    call_xxgetrf(context, builder, xty, x_shapes, o.data, i.data,
+                 info)
+
+    zero = ir.Constant(ll_intc, 0)
+    info_val = builder.load(info)
+    lapack_error = builder.icmp_signed('!=', info_val, zero)
+    invalid_arg = builder.icmp_signed('<', info_val, zero)
+
+    with builder.if_then(lapack_error, False):
+        with builder.if_else(invalid_arg) as (then, otherwise):
+            raise_err = context.call_conv.return_user_exc
+            with then:
+                raise_err(builder, ValueError,
+                          ('One argument passed to getrf is invalid',)
+                          )
+            with otherwise:
+                raise_err(builder, ValueError,
+                          ('Matrix is singular and cannot be inverted',)
+                          )
+
+    # Compute the optimal lwork.
+    lwork = make_constant_slot(context, builder, types.intc, -1)
+    work = cgutils.alloca_once(builder, context.get_value_type(xty.dtype))
+    call_xxgetri(context, builder, xty, x_shapes, o.data, i.data, work,
+                 lwork, info)
+
+    with builder.if_then(lapack_error, False):
+        with builder.if_else(invalid_arg) as (then, otherwise):
+            raise_err = context.call_conv.return_user_exc
+            with then:
+                raise_err(builder, ValueError,
+                          ('One argument passed to getri is invalid',)
+                          )
+            with otherwise:
```

Comment:
Typically, when in workspace query mode, the args are checked for sanity (e.g. check matrix dimensions are positive) then an optimal workspace value is calculated and returned. In this specific case with `xgetri()`, you could for example query the workspace with a singular system in `A` and `info` would be `0` on return as the actual numerical inversion algorithm wouldn't run. Therefore the branch for checking for singularity would never be triggered in this location.

As to the scope of liveness/value held in `info_val` this may be better answered by @pitrou though I somewhat suspect the value needs `builder.load()`ing after each call but will happily be corrected :)